src/openms/source/ANALYSIS/ID/AhoCorasickDA.cpp

```diff
@@ -0,0 +1,820 @@
+      }
+      else
+      {
+        empty_head_ = next;
+      }
+    }
+  }
+
+  void AhoCorasickDA::storeInTail_(const String& suffix)
+  {
+     const char* str = suffix.c_str();
+     while (*str != '\0')
+     {
+       if (isAmbiguous_(*str))
+       {
+         throw Exception::IllegalArgument(__FILE__, __LINE__, OPENMS_PRETTY_FUNCTION, "Input peptide must NOT contain ambiguous amino acids (B/J/Z/X) or characters other than amino acids!");
+       }
+      tail_.push_back(*str++);
+     }
+     tail_.push_back('\0');
+  }
+
+  bool AhoCorasickDA::isAmbiguous_(const char& aa)
+  {
+    //between B (22) and X (25) (B,Z,J,X)
+    return (22 <= code_(aa) && code_(aa) <= 25);
+  }
+
+  void AhoCorasickDA::constructFailure_()
+  {
+    // create empty failure vectors of size bc and tail
+    for (Size i = 0; i < bc_.size(); i++)
+    {
+      failure_bc_.push_back(0);
+    }
+
+    for (Size j = 0; j < tail_.size(); j++)
+    {
+      failure_tail_.push_back(0);
+    }
+
+    std::vector<Int32>nodes;
+    std::vector<Int32>next_nodes;
+
+    //TODO need label (getNode)
+    //edges_ = {'A', 'C', 'D', 'E', 'F', 'G', 'H','I', 'K', 'L', 'M', 'N', 'O', 'P', 'Q', 'R', 'S', 'T', 'U', 'V', 'W', 'Y'};
+
+
+    // compute child nodes of current node_pos
+    for (UInt label = 0; label <= 21; label++)
+    {
+      Int32 node_pos = getNode_(0, label);
+
+      if (node_pos != 0)
+      {
+        nodes.push_back(node_pos);
+      }
+    }
+
+    while(!nodes.empty())
+    {
+
+     for (Int32 node : nodes)
+     {
+       for (UInt label = 0; label <= 21; label++)
+       {
+         Int32 node_pos = getNode_(node, label);
+
+         if (node_pos != 0)
+         {
+           next_nodes.push_back(node_pos);
+           // supply link of parent of node_pos
+           Int32 down = getSupplyLink_(node);
+
+           while ((down != 0) && (getNode_(down, label) == 0)) //
+           {
+             down = getSupplyLink_(down);
+           }
+           // if (down != 0)
+           //{
+           setSupplyLink_(node_pos,getNode_(down, label));
+           //}
+           //else
+           //{
+           // no suffix exists: point suffix link of current node to root
+           //setSupplyLink_(node_pos, 0);
+           //}
+         }
+       }
+     }
+     nodes = next_nodes;
+     next_nodes.clear();
+    }
+  }
+
+  Int32 AhoCorasickDA::getNode_(const Int32& node, const UInt& label)
+  {
+    // node is in tail, child is next index if label match
+    if (node < 0)
+    {
+      if (code_(tail_[abs(node)+1]) == label)
+      {
+        return node -1;
+      }
+      return 0;
+    }
+    // child is tail[base] if label match, returns negative base value because child node is in tail
+    if (bc_[node].leaf_flag)
+    {
+      if (code_(tail_[bc_[node].base]) == label)
+      {
+        return -bc_[node].base;
+      }
+      return 0;
+    }
+
+    // child is in bc
+    else
+    {
+      UInt32 c_node = bc_[node].base ^ label;
```

Comment:
there is a function for this right? This might break if you switch the function, but not the code here.
Always use the function